### PR TITLE
Issue 2

### LINKED ISSUE #2: Self alias should be removed
```C	
typedef struct Foo
{
	int a;
} Foo;
```

Generates:

```D
extern (C):

alias Foo Foo;

struct Foo
{
	int a;
}
```

### PR DESCRIPTION
**Pull request by ciechowoj**
Fixes #2 and #50.

Improves translation of function definitions, they are now broken to multiple lines, if too long and broken in original. It they are single-line in original they are single line in the translation.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Is this PR up to date?

**Comment by ciechowoj** [comment]
Yes.

**Comment by jacob-carlborg** [comment]
https://github.com/jacob-carlborg/dstep/pull/64/commits/9c24ca7f92313776bc5bdb00d7e9debd59517797 seems a bit pointless to have as a separate commit and the commit message seems unrelated.

**Comment by ciechowoj** [comment]
Sorry, it is artifact remaining after rebasing to the `comments` PR and I didn't noticed it...

File: test_files/structs.d [hunk 1]
```diff
@@ -26,28 +26,26 @@ struct A
 
 struct C
 {
-    struct
+    struct _Anonymous_0
     {
         int x;
         int y;
     }
```
Comment (jacob-carlborg) [inline test_files/structs.d]: This doesn't work. Now it's not possible to access `x` or `y`. Ideally it should be translated to:

```d
struct C
{
    struct _Anonymous_0
    {
        int x;
        int y;
    }
    
    _Anonymous_0 point;
}
```

**Comment by jacob-carlborg** [comment]
Does this fix https://github.com/jacob-carlborg/dstep/issues/29 as well?

File: dstep/translator/Output.d [hunk 2]
```diff
@@ -524,9 +691,181 @@ class Output
...
     }
 
+    private Tuple!(size_t, size_t) resolveAdaptiveWidth(size_t itr)
+    {
+        auto data = chunks.data;
+        string[] separators = [ data[itr].separator ];
+        size_t jtr = itr + 1;
+        size_t width = data[itr].content.length;
+
+        size_t separator(size_t jtr, size_t width) {
```
Comment (jacob-carlborg) [inline dstep/translator/Output.d]: Curly brace on the wrong line.

File: test_files/unions.d [hunk 3]
```diff
@@ -26,17 +26,15 @@ union A
 
 union C
 {
-    union
+    union _Anonymous_0
     {
         int x;
         int y;
     }
```
Comment (jacob-carlborg) [inline test_files/unions.d]: Same here as with the struct.

File: dstep/translator/Output.d (hunk 2 above)
Comment (ciechowoj) [inline dstep/translator/Output.d]: Fixed.

File: test_files/structs.d (hunk 1 above)
Comment (ciechowoj) [inline test_files/structs.d]: Fixed, however there is an issue with translation of real anonymous structs (the ones that do no have an associated variable).

File: test_files/unions.d (hunk 3 above)
Comment (ciechowoj) [inline test_files/unions.d]: Fixed, see comment under struct fix.

File: test_files/structs.d (hunk 1 above)
Comment (ciechowoj) [inline test_files/structs.d]: Well, the translation of real anonymous structures is fixed too.

**Comment by ciechowoj** [comment]
> Does this fix #29 as well?

Nope.

File: dstep/translator/Output.d [hunk 4]
```diff
@@ -181,6 +215,112 @@ class Output
         singleLine(fmt, args);
     }
 
+    private Tuple!(string, string, string) adaptiveLineParts(Char, Args...)(
```
Comment (jacob-carlborg) [inline dstep/translator/Output.d]: I find a bit difficult to understand what all these "adaptive" functions are used for. Perhaps some documentation would help.

Comment (ciechowoj) [inline dstep/translator/Output.d]: I'll provide it in a moment.

Comment (ciechowoj) [inline dstep/translator/Output.d]: Done.

**Comment by jacob-carlborg** [comment]
👌

**Comment by ciechowoj** [comment]
> Does this fix #29 as well?
Yes.

File: test_files/enums.d [hunk 5]
```diff
@@ -37,10 +37,11 @@ struct C
     _Anonymous_0 point;
 }
 
-enum _Anonymous_1
+alias i = C._Anonymous_0.i;
+alias j = C._Anonymous_0.j;
```
Comment (jacob-carlborg) [inline test_files/enums.d]: Should this really be aliases at the module level? Aren't the enum members scoped in the struct?

Comment (ciechowoj) [inline test_files/enums.d]: It is problematic:
[C](https://ideone.com/YIkSfE) vs [C++](https://ideone.com/gB6HgD).

Comment (jacob-carlborg) [inline test_files/enums.d]: Aha, didn't know that.

**Comment by jacob-carlborg** [comment]
Seems like this PR is suddenly responsible for handling quite a lot. It now fixes four issues and #39 seems to be completely unrelated to the other ones.

**Comment by ciechowoj** [comment]
Why is this a problem? The features are implemented in self contained commits.

**Comment by jacob-carlborg** [comment]
So far you said the the commits are not self contained. I don't know when a PR is finished you if keep adding more stuff. I need to constantly repeat doing code review and is loosing track of what I already reviewed. And there's no easy way for me to verify that you haven't changed the other commits as well. which I already have reviewed.

Keep the pull requests focused and they should implement/fix one thing.

**Comment by ciechowoj** [comment]
> So far you said the the commits are not self contained.

Yes, but you pointed out that to me on the last PR (`comments`) and now I'm trying to put attention to it.

> I don't know when a PR is finished you if keep adding more stuff.

You didn't accepted it and there was nothing happening here, so I assumed I can add something more (e.g. those enums).

> I need to constantly repeat doing code review and is loosing track of what I already reviewed. And there's no easy way for me to verify that you haven't changed the other commits as well. which I already have reviewed.

I'm very sorry for that.

> Keep the pull requests focused and they should implement/fix one thing.

Sometimes it is hard to do, because things depends on each other, but ok I'll try to do as you say.

The other thing is, I'm GSOC participant and I have quite constrained schedule. Waiting for a review hurts me a lot. I hoped adding more things to one PR will speed up the process...

**Comment by ciechowoj** [comment]
The last three commits are self contained and I haven't changed the former ones since you reviewed it. I'm not sure but I suppose if the commits are overwritten, github moves them at the end of conversation list. The original commits from this PR are still at their initial place.

**Comment by jacob-carlborg** [comment]
> Yes, but you pointed out that to me on the last PR (comments) and now I'm trying to put attention to it.

That's great, I just have to know about it 😃.

> You didn't accepted it and there was nothing happening here, so I assumed I can add something more (e.g. those enums).

I'm still reviewing. I was about to do quick pass through and then accept it but I cannot do that now since you've added new code.

> Sometimes it is hard to do, because things depends on each other, but ok I'll try to do as you say.

I know that can be difficult. There's no ideal solution to this problem, as far as I know. Fortunately Git is a distributed version control system so you should not be too blocked by me lagging behind in the code review. Of course there can be a problem if a code review causes major changes that some other code depends on.

> The other thing is, I'm GSOC participant and I have quite constrained schedule. Waiting for a review hurts me a lot. I hoped adding more things to one PR will speed up the process...

That will certainly _not_  speed up the process. You have to keep in mind that you're working on this, I assume, full time and I'm only working on this on my spare time. Depending on how fast you work and how large changes you do it will be very difficult for me to keep up with your progress.

**Comment by ciechowoj** [comment]
Wonderful : D

**Comment by jacob-carlborg** [comment]
https://travis-ci.org/jacob-carlborg/dstep/jobs/140360262

**Comment by ciechowoj** [comment]
Uff, passed. Seems like some random hiccup.

**Comment by jacob-carlborg** [comment]
Yes, I did rerun the the test. Sorry for the noise.

**Comment by ciechowoj** [comment]
Np.

**Pull request merged**